solr/core/src/java/org/apache/solr/util/circuitbreaker/MemoryCircuitBreaker.java

```diff
@@ -0,0 +1,88 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.util.circuitbreaker;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryMXBean;
+
+import org.apache.solr.core.SolrCore;
+
+public class MemoryCircuitBreaker extends CircuitBreaker {
+  private static final MemoryMXBean MEMORY_MX_BEAN = ManagementFactory.getMemoryMXBean();
+
+  // Assumption -- the value of these parameters will be set correctly before invoking printDebugInfo()
+  private double seenMemory;
+  private double allowedMemory;
+
+  public MemoryCircuitBreaker(SolrCore solrCore) {
+    super(solrCore);
+  }
+
+  // TODO: An optimization can be to trip the circuit breaker for a duration of time
+  // after the circuit breaker condition is matched. This will optimize for per call
+  // overhead of calculating the condition parameters but can result in false positives.
+  @Override
+  public boolean isCircuitBreakerGauntletTripped() {
+    if (!isCircuitBreakerEnabled()) {
+      return false;
+    }
+
+    allowedMemory = getCurrentMemoryThreshold();
+
+    if (allowedMemory < 0) {
+      // No threshold
+      return false;
+    }
+
+    seenMemory = calculateLiveMemoryUsage();
+
+    return (seenMemory >= allowedMemory);
+  }
+
+  @Override
+  public String printDebugInfo() {
+    return "seen memory " + seenMemory + " allowed memory " + allowedMemory;
+  }
+
+  private double getCurrentMemoryThreshold() {
+    int thresholdValueInPercentage = solrCore.getSolrConfig().memoryCircuitBreakerThreshold;
+    long currentMaxHeap = MEMORY_MX_BEAN.getHeapMemoryUsage().getMax();
+
+    if (currentMaxHeap <= 0) {
+      return Long.MIN_VALUE;
+    }
+
+    double thresholdInFraction = (double) thresholdValueInPercentage / 100;
+    double actualLimit = currentMaxHeap * thresholdInFraction;
+
+    if (actualLimit <= 0) {
+      throw new IllegalStateException("Memory limit cannot be less than or equal to zero");
+    }
+
+    return actualLimit;
+  }
+
+  /**
+   * Calculate the live memory usage for the system. This method has package visibility
+   * to allow using for testing
+   * @return Memory usage in bytes
+   */
+  protected long calculateLiveMemoryUsage() {
+    return MEMORY_MX_BEAN.getHeapMemoryUsage().getUsed();
```

Comment:
Is this expensive to get? If we're doing it for every query, I'd want to make sure that it is an efficient call.

Reply:
I did not see any regression -- if we note anything otherwise, we can probably cache these values with a TTL. I would be surprised if it causes anything noticeable, though.

Reply:
This article claims that each call is about 100ns, https://www.inoneo.com/en/blog/8/java/monitoring-jvm-memory-usage-inside-a-java-application and suggest using a listener instead. I'm not sure how true this is for more modern JVMs anyway.

If you look into drop wizard metrics that we're reporting in JMX, MemoryUsageGuageSet reports memory as the combination of heap and non-heap. I'm not sure we need to consider non-heap here, but at least a comment highlighting that would be useful.

Reply:
Yes, I considered MemoryUsageGaugeSet and did not use it for the precise reason that you described. To be honest, the listener based implementation is harder to read than the current, and I was not sure if it is worth going that route for shaving off < 100 ns per query. However, if you have concerns or if we ever hear slowness reports due to this feature, I am happy to take that route. For now, I have added the comment as suggested, thanks.

Reply:
@madrob Just verified that Elasticsearch also uses getHeapMemoryUsage() to trigger similar checks per call. Also, I am not sure if NotificationListener will allow us to be real time since it triggers a notification when the threshold is breached. but I could not find a way to know when the threshold has gone back to normal. I am proposing sticking to this for now, if that sounds fine?

Reply:
Yea, add some comments to this effect to the code so that if people run into issues with it they can more easily figure out what's up without having to find the comments from this PR.

Reply:
Added a comment, thanks